plugins/com.google.cloud.tools.eclipse.ui.status/src/com/google/cloud/tools/eclipse/ui/status/PollingStatusServiceImpl.java

```diff
@@ -0,0 +1,196 @@
+import java.net.URI;
+import java.net.URLConnection;
+import java.nio.charset.StandardCharsets;
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.logging.Logger;
+import org.eclipse.core.net.proxy.IProxyData;
+import org.eclipse.core.net.proxy.IProxyService;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.ListenerList;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.core.runtime.jobs.Job;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+
+/**
+ * An implementation that polls the Google Cloud Platform's status page. The Google Cloud Platform
+ * status page provides a incident log in JSON, which appears to be ordered from most recent to
+ * oldest. We fetch the first N bytes and process the incidents listed. Incidents that are still
+ * on-going do not have an "end".
+ */
+@Component(name = "polling")
+public class PollingStatusServiceImpl implements GcpStatusService {
+  private static final Logger logger = Logger.getLogger(PollingStatusServiceImpl.class.getName());
+
+  private static final URI STATUS_JSON_URI =
+      URI.create("https://status.cloud.google.com/incidents.json");
+
+  private Job pollingJob =
+      new Job("Retrieving Google Cloud Platform status") {
+        @Override
+        protected IStatus run(IProgressMonitor monitor) {
+          refreshStatus();
+          if (active) {
+            schedule(pollTime);
+          }
+          return Status.OK_STATUS;
+        }
+      };
+
+  private boolean active = false;
+  private long pollTime = 3 * 60 * 1000; // poll every 3 minutes
+  private IProxyService proxyService;
+  private ListenerList /*<Consumer<GcpStatusService>>*/ listeners = new ListenerList /*<>*/();
+  private Gson gson = new Gson();
+
+  private GcpStatus currentStatus = GcpStatus.OK_STATUS;
+
+  @Reference(policy = ReferencePolicy.DYNAMIC, cardinality = ReferenceCardinality.OPTIONAL)
+  public void setProxyService(IProxyService proxyService) {
+    this.proxyService = proxyService;
+  }
+
+  public void unsetProxyService(IProxyService proxyService) {
+    if (this.proxyService == proxyService) {
+      this.proxyService = null;
+    }
+  }
+
+  @Activate
+  public void start() {
+    active = true;
+    pollingJob.schedule();
+  }
+
+  @Deactivate
+  public void stop() {
+    active = false;
+    pollingJob.cancel();
+  }
+
+  @Override
+  public GcpStatus getCurrentStatus() {
+    return currentStatus;
+  }
+
+  void refreshStatus() {
+    try {
+      // As of 2018-01-30 the incidents log is 258k! But the incidents appear to be sorted from most
```

Comment:
Is this documented anywhere? Can we count on this remaining available at this URL?

Reply:
No, it doesn't seem to be documented anywhere that I can see.  It seems to be regenerated about once a minute, judging from the `Last-Modified` headers.

If I switch to the Google HTTP client code, then we should be able to fetch the entire file with gzip encoding which reduces it to 36k. That would fetch approx 720kb/hour at a 3-minute refresh interval.